Add CRL checking for client certificate

Vagabond opened the pull request:
If the client provides a certificate, make sure it is actually valid AND
not revoked.

See also basho/riak_test#416

Vagabond committed:
Add CRL checking for client certificate

If the client provides a certificate, make sure it is actually valid AND
not revoked.

Vagabond commented:
Oh, for this to actually work, you'll need to patch public_key:

https://github.com/erlang/otp/pull/97

cmeiklejohn commented on src/riak_api_pb_server.erl:
```diff
@@ -520,4 +523,172 @@ send_error_and_flush(Error, State) ->
+    fetch(Rest);
+fetch([{uniformResourceIdentifier, "http"++_=URL}|Rest]) ->
+    lager:debug("getting CRL from ~p~n", [URL]),
+    inets:start(),
```
Do we have to worry about an error being returned here because it's already been started?

cmeiklejohn commented:
I don't mean to be overly pedantic here, but adding some docstrings to the functions would go a long way to being able to revisit this code later and know why it does certain things.

cmeiklejohn commented on src/riak_api_pb_server.erl:
```diff
@@ -520,4 +523,172 @@ send_error_and_flush(Error, State) ->
+    %% XXX the 'Issuer' we get passed here is actually a lie, public key treats the Authority Key Identifier as the 'issuer'
+    %% Read the CA certs out of the file
+    %{ok, Bin} = file:read_file(CACerts),
+    %CACerts = load_certs("/home/andrew/lavabit"),
```
These commented lines probably shouldn't be here.

cmeiklejohn commented on src/riak_api_pb_server.erl:
```diff
@@ -520,4 +523,172 @@ send_error_and_flush(Error, State) ->
+      end, undefined, Certs).
+
+fetch_point(#'DistributionPoint'{distributionPoint = P}) ->
+    case P of
```
Is this really necessary?  Seems like you could just explicitly match P in the function clause?

cmeiklejohn committed:
Reformatting.

Add comments; verify clean Dialyzer, test against Riak Test, and
document SSL functions.

cmeiklejohn commented:
:+1: once #40 is merged.  Should address most of my comments in this PR.

Vagabond committed:
Merge pull request #40 from basho/bugfix/csm/reformatting-adt-pb-crl

Reformatting.
